File: 0611-valid-triangle-number/0611-valid-triangle-number.cpp
class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        int count=0;
        if(nums.size()<3)
            return 0;
        sort(nums.begin(),nums.end());
        int i=0;
        while(nums[i]==0)
            ++i;
        for( ;i<nums.size()-2;++i)
        {
            for(int j=i+1;j<nums.size()-1;++j)
            {
                int l=j+1,h=nums.size()-1,mid,k=nums[i]+nums[j];
                while(l<=h)
                {
                    mid=l+(h-l)/2;
                    if(nums[mid]>=k)
                        h=mid-1;
                    else 
                        l=mid+1;
                }
                count=count+(h-j);
            }
        }
        return count;
    }
};

File: 0611-valid-triangle-number/0611-valid-triangle-number.py
class Solution:
    def calc(self,sum,l,nums):
        h=len(nums)-1
        ans = l-1
        while l<=h:
            mid= h-(h-l)//2
            if nums[mid]<sum:
                ans = mid
                l = mid+1
            else:
                h=mid-1
        #print(ans)
        return ans
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        ans =0
        n = len(nums)
        for i in range(n):
            for j in range(i+1,n):
                ans += self.calc(nums[i]+nums[j],j+1,nums)-j
        return ans
        